File: gdx2d-demoDesktop/src/ch/hevs/gdx2d/demos/physics/DemoPolygonPhysics.java
package ch.hevs.gdx2d.demos.physics;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.math.Vector2;
import hevs.gdx2d.components.graphics.GeomUtils;
import hevs.gdx2d.components.physics.primitives.PhysicsPolygon;
import hevs.gdx2d.components.physics.utils.PhysicsScreenBoundaries;
import hevs.gdx2d.lib.GdxGraphics;
import hevs.gdx2d.lib.PortableApplication;
import hevs.gdx2d.lib.physics.DebugRenderer;
import hevs.gdx2d.lib.physics.PhysicsWorld;

/**
 * Demonstrates how to use general polygons as physics objects
 *
 * @author Pierre-André Mudry
 * @version 1.0
 */
public class DemoPolygonPhysics extends PortableApplication {

	PhysicsPolygon p1, p2;
	DebugRenderer debugRenderer;

	public static void main(String[] args) {
		new DemoPolygonPhysics();
	}

	@Override
	public void onInit() {
		// A triangle
		Vector2 obj1[] = {
				new Vector2(100, 100),
				new Vector2(200, 100),
				new Vector2(150, 200)
		};

		// A special polygon
		Vector2 obj2[] = {
				new Vector2(0, 0),
				new Vector2(4, 0),
				new Vector2(5, 3),
				new Vector2(2, 8),
				new Vector2(-1, 3)
		};

		GeomUtils.translate(obj1, new Vector2(250, 100));
		GeomUtils.rotate(obj1, 12);

		GeomUtils.scale(obj2, 20);
		GeomUtils.rotate(obj2, 14);
		GeomUtils.translate(obj2, new Vector2(100, 100));

		p1 = new PhysicsPolygon("poly1", obj1, true);
		p2 = new PhysicsPolygon("poly2", obj2, true);

		new PhysicsScreenBoundaries(this.getWindowWidth(), this.getWindowHeight());

		debugRenderer = new DebugRenderer();
	}

	@Override
	public void onGraphicRender(GdxGraphics g) {
		g.clear();

		PhysicsWorld.updatePhysics(Gdx.graphics.getRawDeltaTime());
		debugRenderer.render(PhysicsWorld.getInstance(), g.getCamera().combined);

		g.drawFilledPolygon(p1.getPolygon(), Color.YELLOW);
		g.drawFilledPolygon(p2.getPolygon(), Color.RED);

		g.drawSchoolLogo();
		g.drawFPS();
	}

}

File: gdx2d-demoDesktop/src/main/scala/ch/hevs/gdx2d/demos/physics/DemoPolygonPhysics.scala
package ch.hevs.gdx2d.demos.physics

import ch.hevs.gdx2d.components.graphics.GeomUtils
import ch.hevs.gdx2d.components.physics.primitives.PhysicsPolygon
import ch.hevs.gdx2d.components.physics.utils.PhysicsScreenBoundaries
import ch.hevs.gdx2d.desktop.DesktopApplication
import ch.hevs.gdx2d.desktop.physics.DebugRenderer
import ch.hevs.gdx2d.lib.GdxGraphics
import ch.hevs.gdx2d.lib.physics.PhysicsWorld
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.math.Vector2

/**
 * Demonstrates how to use general polygons as physics objects.
 *
 * @author Pierre-André Mudry
 * @version 2.0
 */
class DemoPolygonPhysics extends DesktopApplication {

  private var p1: PhysicsPolygon = _
  private var p2: PhysicsPolygon = _
  private var debugRenderer: DebugRenderer = _

  override def onInit(): Unit = {
    // A triangle
    val obj1 = Array(new Vector2(100f, 100f), new Vector2(200f, 100f), new Vector2(150f, 200f))

    // A special polygon
    val obj2 = Array(new Vector2(0f, 0f), new Vector2(4f, 0f), new Vector2(5f, 3f), new Vector2(2f, 8f), new Vector2(-1f, 3f))

    GeomUtils.translate(obj1, new Vector2(250f, 100f))
    GeomUtils.rotate(obj1, 12f)

    GeomUtils.scale(obj2, 20f)
    GeomUtils.rotate(obj2, 14f)
    GeomUtils.translate(obj2, new Vector2(100f, 100f))

    p1 = new PhysicsPolygon("poly1", obj1, true)
    p2 = new PhysicsPolygon("poly2", obj2, true)

    new PhysicsScreenBoundaries(getWindowWidth.toFloat, getWindowHeight.toFloat)

    debugRenderer = new DebugRenderer()
  }

  override def onGraphicRender(g: GdxGraphics): Unit = {
    g.clear()

    PhysicsWorld.updatePhysics(Gdx.graphics.getRawDeltaTime)
    debugRenderer.render(PhysicsWorld.getInstance(), g.getCamera.combined)

    g.drawFilledPolygon(p1.getPolygon, Color.YELLOW)
    g.drawFilledPolygon(p2.getPolygon, Color.RED)

    g.drawSchoolLogo()
    g.drawFPS()
  }
}

object DemoPolygonPhysics {
  def main(args: Array[String]): Unit = {
    new DemoPolygonPhysics().launch()
  }
}
